types/index.d.ts

```diff
@@ -0,0 +1,215 @@
+  ssl?: boolean | import('tls').TlsOptions;
+  /** Max number of connections */
+  max?: number;
+  /** Idle connection timeout in seconds */
+  timeout?: number;
+  /** Array of custom types; see more below */
+  types?: T;
+  /** Defaults to console.log */
+  onnotice?: (notice: Postgres.Notice) => void;
+  /** (key; value) when server param change */
+  onparameter?: (key: string, value: any) => void;
+  /** Is called with (connection; query; parameters) */
+  debug?: (connection: number, query: string, parameters: any[]) => void;
+  /** Transform hooks */
+  transform?: {
+    /** Transforms incoming column names */
+    column?: (column: string) => any;
+    /** Transforms incoming row values */
+    value?: (value: any) => any;
+    /** Transforms entire rows */
+    row?: (row: Postgres.Row) => any;
+  };
+  /** Connection parameters */
+  connection?: Postgres.ConnectionVariables;
+}
+
+type UnwrapPromiseArray<T> = T extends any[] ? {
+  [k in keyof T]: T[k] extends Promise<infer R> ? R : T[k]
+} : T;
+
+declare namespace Postgres {
+
+  /**
+   * Convert a string to Pascal case.
+   * @param str THe string to convert
+   * @returns The new string in Pascal case
+   */
+  function toPascal(str: string): string;
+  /**
+   * Convert a string to Camel case.
+   * @param str THe string to convert
+   * @returns The new string in Camel case
+   */
+  function toCamel(str: string): string;
+  /**
+   * Convert a string to Kebab case.
+   * @param str THe string to convert
+   * @returns The new string in Kebab case
+   */
+  function toKebab(str: string): string;
+
+  interface ConnectionVariables {
+    /** Default application_name */
+    application_name?: string;
+    /** Other connection parameters */
+    [name: string]: any;
+  }
+
+  interface Options<T extends CustomTypeList> extends BaseOptions<T> {
+    /** unix socket path (usually '/tmp') */
+    path?: string;
+    /** Password of database user */
+    pass?: string | (() => string | Promise<string>);
+    /** Password of database user */
+    password?: Options<T>['pass']; // FIXME Is it a doc error ?
+  }
+
+  interface ParsedOptions<T extends CustomTypeList> extends BaseOptions<T> {
+    /** @inheritdoc */
+    port: number;
+    /** @inheritdoc */
+    path: string | false;
+    /** @inheritdoc */
+    database: string;
+    /** @inheritdoc */
+    user: string;
+    /** @inheritdoc */
+    pass: null;
+    /** @inheritdoc */
+    max: number;
+    /** @inheritdoc */
+    types: T;
+    /** @inheritdoc */
+    transform: NonNullable<Options<T>['transform']>
+    /** @inheritdoc */
+    ssl: boolean;
+    /** @inheritdoc */
+    serializers: { [oid: number]: CustomType['serialize'] };
+    /** @inheritdoc */
+    parsers: { [oid: number]: CustomType['parse'] };
+  }
+
+  interface Notice {
+    [field: string]: string;
+  }
+
+  interface CustomType {
+    to: number,
+    from: number[],
+    serialize(obj: unknown): unknown,
+    parse(raw: any): unknown
+  }
```

Comment:
Wouldn't type-parameter make more sense here?
```typescript
  interface CustomType<T> {
    to: number
    from: number[]
    serialize(obj: T): string
    parse(raw: string): T
  }
```

Reply:
Not really, as the `T` type-parameter would have to be passed everywhere it is used, and moreover multiple types can be specified in custom types list, so a commun `T` for all is weak.

Instead of using a type-parameter, I let TypeScript infer the whole custom types list https://github.com/porsager/postgres/blob/4e5427086cea596de5a640813b1a95402fa727e2/types/index.d.ts#L133-L135
so that I can use it as a mapping type : https://github.com/porsager/postgres/blob/4e5427086cea596de5a640813b1a95402fa727e2/types/index.d.ts#L202-L204

Reply:
It's the definition of how to handle specific postgres types, so maybe simply `PostgresType`?

Reply:
Should I rename `TTypes`, `CustomType` or `CustomTypeList` ?

Reply:
No I think it's fine if you feel good about them, but if you think `PostgresType`, `PostgresTypeList` make better sense then feel free :)

I'm not sure about `TTypes`. If I understand correctly it's not something a typescript user would be exposed to, so it's not that important?

Reply:
Normally I would have called it `T` but it is already used by functions in `Sql`, so I chose `TTypes`. Indeed, It is not exposed but can help to understand the file.

Reply:
> Instead of using a type-parameter, I let TypeScript infer the whole custom types list

Yes, I got your intention. But may be we could push `unknown` a little bit higher:
```typescript
 interface CustomTypeList { 
   [name: string]: CustomType<unknown, unknown>
 }
```
and the compiler will at least type-check that `serialize` and `parse` functions does match each other? Not sure about that though...